Add UMAP to embedding projector

cannoneyed opened the pull request:
* Motivation for features / changes

Uniform Manifold Approximation and Projection [(UMAP)](https://github.com/lmcinnes/umap) is a dimension reduction technique that can be used for visualisation similarly to t-SNE. I've ported the python implementation to JavaScript to add an additional projection method to the embedding projector.

* Technical description of changes

The UMAP projection technique pretty much follows the approach for the tSNE projection. The UMAP code is grouped in its own directory for ease of maintenance. There are two things to note about this current implementation of UMAP in the embedding projector:
1) UMAP's SGD optimization doesn't really animate as nicely as t-SNE. I've addresses this by choosing not to animate the steps and just track progress through a dialog.
1) I'm currently reusing the GPU computed k-nearest neighbors - there didn't seem to be too big of a difference between using the UMAP CPU knn approach vs the existing GPU approach in the projector. To note, the GPU knn approach uses cosine distance, while the rest of UMAP's implementation uses euclidean distance. Again, the subjective results seem fine with this choice.

* Screenshots of UI changes

![image](https://user-images.githubusercontent.com/7818584/52659999-88c2c400-2eb3-11e9-9b68-eab380cbf11d.png)

* Detailed steps to verify changes work correctly (as executed by you)


* Alternate designs / implementations considered

wchargin commented:
Hi @cannoneyed! Thanks for the PR. We’ll likely be interested in
integrating this functionality after the TF Dev Summit next week, for
which we’re currently busy preparing.

cannoneyed commented:
Hey William! Sounds great, I figured I'd get everything in-place then we can move forward once you're through tf-dev.

wchargin commented on tensorboard/plugins/projector/vz_projector/data.ts:
```diff
@@ -370,6 +378,84 @@ export class DataSet {
   }
 
+  /** Runs UMAP on the data. */
+  projectUmap(nComponents: number, nNeighbors: number, stepCallback: (iter: number) => void) {
```
So, after reading through this logic, the control flow still isn’t clear
to me. Why is `requestAnimationFrame` only called once, outside of
`step`, while `step` itself recurs synchronously? The `nEpochs`
variable, which appears to hold the target (maximum) number of epochs,
is initialized to `1`, but later mutated from within a _different_
asynchronous callback to the result of `initializeFit`, so which value
is used when?

Is there some way that we can restructure this to make the control flow
obvious from the structure of the code? For instance, if the initial
value of `1` for `nEpochs` is supposed to be an unobservable state, then
could we write something like

```typescript
async projectUmap(/* … */) {
  const nearest = await /* compute knnComputation */;
  const nEpochs = await util.runAsyncTask('Initializing UMAP…', () => {
    // …
    return this.umap.initializeFit(/* … */);
  });
  return new Promise((resolve) => {
    const step = () => {
      // …
      requestAnimationFrame(step);
      // …
      if (done) resolve();
    };
    requestAnimationFrame(step);
  });
}
```

(I think that I’ve reached my short-term saturation on understanding
this function and allotting time to it, so I’ll come back to see what
you come up with. :-) )

cannoneyed replied:
Great, thanks for such a thorough review.

Yeah, I wasn't particularly pleased with how I wound up doing this, it was pretty much hacked together to work like the t-SNE implementation before I had a really good grasp on how the application worked together. Anyway, I've refactored it completely to make it a bit more clear what's going on and why it's going on that way.

wchargin replied:
Thank you! This is much easier for me to read.

cannoneyed commented on third_party/js.bzl:
```diff
@@ -206,6 +206,22 @@ def tensorboard_js_workspace():
   )
 
+  filegroup_external(
+      name = "io_github_PAIR-code_umap-js",
```
👍

cannoneyed commented on tensorboard/plugins/projector/vz_projector/vz-projector-projections-panel.ts:
```diff
@@ -132,7 +141,10 @@ export class ProjectionsPanel extends ProjectionsPanelPolymer {
+    this.iterationLabelTSNE = this.querySelector('.run-tsne-iter') as HTMLElement;
+    this.umapNeighborsInput =
+        this.querySelector('#umap-neighbors-slider') as HTMLInputElement;
+    this.runUmapButton = this.querySelector('.run-umap') as HTMLButtonElement;
```
Nope. Just following the (confusing and inconsistent) patterns from the rest of the app... I think it might be worth a full-pass comb-through of a bunch of these inconsistencies around how the polymer part of the app is implemented, and I'd be happy to do that in another PR

cannoneyed commented on tensorboard/plugins/projector/vz_projector/vz-projector-projections-panel.ts:
```diff
@@ -167,6 +179,14 @@ export class ProjectionsPanel extends ProjectionsPanelPolymer {
 
+  private updateUmapNeighborsFromUIChange() {
+    if (this.umapNeighborsInput) {
+      this.umapNeighbors = + this.umapNeighborsInput.value;
```
Yeah, sounds great! Apologies for being a complete polymer noob - I basically just copy-pasted logic from similar components of the t-SNE controls. But happy to do it with less manual listener nonsense.

cannoneyed commented on tensorboard/plugins/projector/vz_projector/vz-projector-projections-panel.ts:
```diff
@@ -497,6 +537,51 @@ export class ProjectionsPanel extends ProjectionsPanelPolymer {
+    this.runUmapButton.disabled = true;
+
+    const nComponents = this.umapIs3d ? 3 : 2;
+    const nNeighbors = this.umapNeighbors
```
Is there a `clang` style code formatter for this project?

cannoneyed commented:
OK, I've implemented the requested changes and refactored the `projectUmap` method to make that control flow a bit more clear. Thanks for working through this so thoroughly!

wchargin commented on tensorboard/plugins/projector/vz_projector/data.ts:
```diff
@@ -370,6 +366,92 @@ export class DataSet {
+      // We found the nearest neighbors before and will reuse them.
+      return Promise.resolve(this.nearest);
+    } else {
+      this.nearestK = nNeighbors;
```
Also, I don’t really understand why `this.nearestK` needs to be
persisted at all. Isn’t it always `this.nearest.length` (and likewise
undefined when `this.nearest == null`)?

cannoneyed replied:
Yeah, that's a good catch. And I agree with you in regards to the `nearestK` point - just following patterns blindly y'know? 😛

cannoneyed commented on tensorboard/plugins/projector/vz_projector/data.ts:
```diff
@@ -370,6 +366,92 @@ export class DataSet {
+
+    this.nearest = await this.computeKnn(sampledData, nNeighbors);
+    
+    nEpochs = await util.runAsyncTask('Initializing UMAP...', () => {
```
Good catch

cannoneyed commented on tensorboard/plugins/projector/vz_projector/data.ts:
```diff
@@ -370,6 +366,92 @@ export class DataSet {
+
+    this.nearest = await this.computeKnn(sampledData, nNeighbors);
+    
+    nEpochs = await util.runAsyncTask('Initializing UMAP...', () => {
```
The initialization takes a while, maybe 10-20s for the demo dataset (n=5000, d=784)... There are a few steps being performed in this phase, and I've thought about later updating umap-js to expose the progress in these steps via a callback to display amore helpful message in the UI, but I thought this .

The `initializeFit` function is synchronous, and the promise returned by `util.runAsyncTask` is resolved after it is completed. I think that this variant of the helper function was originally created to "buffer" computationally expensive synchronous actions (like updating the positions of points in the t-SNE optimization steps) with some timeout time around them, in order to prevent the UX from freezing up during these steps. 

Are you saying I should use the more generic `tf_graph_common/util.ts` helper methods here? The only bummer that I see with that is the embedding projector doesn't conform to the `ProgressTracker` API in the generic util.

cannoneyed commented on tensorboard/plugins/projector/vz_projector/data.ts:
```diff
@@ -370,6 +366,92 @@ export class DataSet {
+        // Wrap the logic in a util.runAsyncTask in order to correctly update
+        // the modal with the progress of the optimization.
+        util.runAsyncTask(progressMsg, () => {
+        }, UMAP_MSG_ID, 0).then(() => {
```
It's effectively the same thing (`runAsyncTask(noop).then(work, errorCB)` vs `runAsyncTask(work).then(noop, errorCB)`) and I had just copied over the success/error callback structure  from the t-SNE implementation. I've refactored to put the work in the `runAsyncTask` callback and use a `catch` call instead of the success/error pattern, which I think is now more clear.

cannoneyed commented on tensorboard/plugins/projector/vz_projector/data.ts:
```diff
@@ -370,6 +366,92 @@ export class DataSet {
+    // the UMAP animation is not nearly as informative as t-SNE.
+    return new Promise((resolve, reject) => {
+      const step = () => {
+        // Compute a batch of epochs since we don't want to update the UI
```
Correct, my suspicion is that the way I initialize the optimization step in the JS version (random) vs the default in python (spectral graph cut) means that the animation looks pretty awful while optimizing UMAP (very "thrashy")... until I get a chance to implement the spectral initialization (which is a pretty big can of worms), then I think the best option is just to update progress in the modal...

We might want to think about adding a "cancel" button to the modal, or just make the experience of starting the projection a bit better (right now switching projection tabs automatically starts the process, which is a bit annoying for UMAP where there's no cancel button)

cannoneyed commented on tensorboard/plugins/projector/vz_projector/vz-projector-projections-panel.ts:
```diff
@@ -97,14 +105,18 @@ export class ProjectionsPanel extends ProjectionsPanelPolymer {
   private customProjectionYDownInput: ProjectorInput;
 
   initialize(projector: Projector) {
+    console.log(this);
```
Thanks!

cannoneyed commented on tensorboard/plugins/projector/vz_projector/vz-projector-projections-panel.ts:
```diff
@@ -497,6 +523,51 @@ export class ProjectionsPanel extends ProjectionsPanelPolymer {
+        new Projection('umap', accessors, dimensionality, dataSet);
+    this.projector.setProjection(projection);
+
+    if (!this.dataSet.hasTSNERun) {
```
Thanks again! As far as the PCA/t-SNE issue, I've seen this and haven't yet figured out what the cause is... Maybe I can fix it in some subsequent work?

cannoneyed commented:
Thanks for getting around to this, and hopefully we can push through the finish line soon! After seeing how much work you've put in / how much work could be done for this and subsequent PRs, I think it's worth revisiting the amount of support that I can provide for the projector. Let me talk with my team a bit to see how much time I can reasonably expect to put into supporting / updating the projector in the near term!

wchargin commented:
Sounds good; thanks.

wchargin commented:
You can rebase master (which now includes #2052) to drop the cherry-pick
that Googlebot is complaining about.

wchargin commented on tensorboard/plugins/projector/vz_projector/util.ts:
```diff
@@ -248,4 +250,9 @@ export function hasWebGLSupport(): boolean {
 }
 
+/** Checks to see if input is undefined or null. */
+export function exists(x: any): boolean {
```
IMO this doesn’t merit its own function. The pattern `x != null` is
sufficiently common in JavaScript/TypeScript (192 uses in our codebase)
that when reading `util.exists` I assume that this function must be
doing something _else_ and have to look it up to see that in fact it is
doing the same thing.

Although <http://go/tsstyle> (Google-internal) bans `!=` in general, the
particular case of `!= null` is explicitly permitted.

cannoneyed replied:
Fine with me! Though I do have an irrational fear of coercion...

wchargin commented on tensorboard/plugins/projector/vz_projector/data.ts:
```diff
@@ -370,6 +365,91 @@ export class DataSet {
+          } else {
+            const result = this.umap.getEmbedding();
+            sampledIndices.forEach((index, i) => {
+               const dataPoint = this.points[index];
```
nit: extra leading space

cannoneyed replied:
Thx! (By the way, have you ever used `[prettier](https://github.com/prettier/prettier)`? Do you know how easy it is to get autoformatting working w bazel?)

wchargin commented on tensorboard/plugins/projector/vz_projector/data.ts:
```diff
@@ -370,6 +365,91 @@ export class DataSet {
+  }  
+
+  /** Computes KNN to provide to the UMAP and t-SNE algorithms. */
+  private async computeKnn(data: DataPoint[], nNeighbors: number): Promise<knn.NearestEntry[][]> {
```
nit: Please wrap at 80 like the rest of this file; Prettier suggests:

```typescript
  private async computeKnn(
    data: DataPoint[],
    nNeighbors: number,
  ): Promise<knn.NearestEntry[][]> {
    /* … */
  }
```

cannoneyed replied:
I'm basically useless without prettier.

wchargin commented on tensorboard/plugins/projector/vz_projector/data.ts:
```diff
@@ -370,6 +365,91 @@ export class DataSet {
+    
+    const nEpochs = await util.runAsyncTask('Initializing UMAP...', () => {
+      const knnIndices = this.nearest.map(row => row.map(entry => entry.index));
+      const knnDistances = this.nearest.map(row => row.map(entry => entry.dist));
```
nit: Please wrap at 80 like the rest of this file; Prettier suggests:

```typescript
      const knnDistances = this.nearest.map((row) =>
        row.map((entry) => entry.dist),
      );
```

cannoneyed replied:
oh, answered my question re: prettier

wchargin commented on tensorboard/plugins/projector/vz_projector/data.ts:
```diff
@@ -370,6 +365,91 @@ export class DataSet {
+        for (let i = 0; i < epochsBatch; i++) {
+          currentEpoch = this.umap.step();
+        }
+        const progressMsg = `Optimizing UMAP (epoch ${currentEpoch} of ${nEpochs})`;
```
nit: Please wrap at 80 like the rest of this file; suggest:

```typescript
        const progressMsg =
          `Optimizing UMAP (epoch ${currentEpoch} of ${nEpochs})`;
```

wchargin commented on tensorboard/plugins/projector/vz_projector/vz-projector-projections-panel.ts:
```diff
@@ -497,6 +521,51 @@ export class ProjectionsPanel extends ProjectionsPanelPolymer {
+              projectionChangeNotified = true;
+            }
+          }
+          else {
```
nit: `else` goes on the same line as the preceding closing brace.

cannoneyed commented:
Thanks for taking the time for such a thorough review. What's the timeline look like for deployment once these changes are merged?

wchargin commented:
> What's the timeline look like for deployment once these changes are
> merged?

I’ll merge this into master, so it will be in `tb-nightly` starting
tonight. It’s a new feature, so it will be generally available in the
next minor version of TensorBoard: namely, TensorBoard 1.14.0, which
will roughly coincide with the TensorFlow 1.14.0 release.
